### PR TITLE
use fee swap path

### PR DESCRIPTION
**Pull request by wangjj9219**

### CODE REVIEW

File: modules/transaction-payment/src/lib.rs [hunk 1]
```diff
@@ -355,19 +364,33 @@ pub mod module {
 				);
 			})
 		}
+
+		// remove after runtime upgraded
+		fn on_runtime_upgrade() -> Weight {
+			DefaultFeeCurrencyId::<T>::remove_all(None);
+			T::DbWeight::get().reads_writes(0, 1)
```
Comment (xlc) [inline modules/transaction-payment/src/lib.rs:371]: we need to check how many items we have currently and change writes.

**Review by xlc — APPROVED** [review]

File: modules/transaction-payment/src/lib.rs (hunk 1 above)
Comment (wangjj9219) [reply modules/transaction-payment/src/lib.rs]: There are currently 34 items in total, which is an easy mission for upgrades.

File: modules/transaction-payment/src/lib.rs [hunk 2]
```diff
@@ -573,38 +587,46 @@ where
...
-				let trading_path = if currency_id == stable_currency_id {
-					vec![stable_currency_id, native_currency_id]
+		// check native balance if is enough
+		let native_is_enough = <T as Config>::Currency::free_balance(who)
+			.checked_sub(&fee_and_alive_gap)
+			.map_or(false, |new_free_balance| {
+				<T as Config>::Currency::ensure_can_withdraw(who, fee, reason, new_free_balance).is_ok()
+			});
+
+		// native is enough, try swap native to pay fee
```
Comment (xlc) [inline modules/transaction-payment/src/lib.rs:597]:
```suggestion
		// native is not enough, try swap native to pay fee
```

File: modules/transaction-payment/src/lib.rs [hunk 3]
```diff
@@ -573,38 +587,46 @@ where
...
-					&trading_path,
-					fee_and_alive_gap.unique_saturated_into(),
-					<T as Config>::MultiCurrency::free_balance(currency_id, who),
-					price_impact_limit,
-				)
-				.is_ok()
-				{
-					// successfully swap, break iteration
-					break;
+			fee_swap_path_list.dedup();
```
Comment (xlc) [inline modules/transaction-payment/src/lib.rs:608]: dedup only works for sorted vec

File: modules/transaction-payment/src/lib.rs [hunk 4]
```diff
@@ -284,20 +280,33 @@ pub mod module {
...
 	/// DefaultFeeCurrencyId: AccountId => Option<CurrencyId>
 	#[pallet::storage]
 	#[pallet::getter(fn default_fee_currency_id)]
 	pub type DefaultFeeCurrencyId<T: Config> = StorageMap<_, Twox64Concat, T::AccountId, CurrencyId, OptionQuery>;
 
+	/// The alternative fee swap path of accounts.
+	#[pallet::storage]
+	#[pallet::getter(fn alternative_fee_swap_path)]
+	pub type AlternativeFeeSwapPath<T: Config> =
+		StorageMap<_, Twox64Concat, T::AccountId, Vec<CurrencyId>, OptionQuery>;
```
Comment (shaunxw) [inline modules/transaction-payment/src/lib.rs:308]: `Twox64Concat` might not be safe enough for `T::AccountId`?

For swap path, we might need bounded vec instead.

Comment (xlc) [reply modules/transaction-payment/src/lib.rs:308]: People can only set swap path for themself so it should be safe.
Can hardcode the limit to 3 elements for swap path.

**Comment by shaunxw** [comment]
How about the `Twox64Concat` hasher used for `T::AccountId`, it should be changed to blake2 ...no?

**Comment by xlc** [comment]
It is safe. Only account owner can set that.

**Pull request merged**